src/group_theory/congruence.lean

```diff
@@ -419,5 +429,342 @@ protected def gi : @galois_insertion (M → M → Prop) (con M) _ _ con_gen r :=
+  one_mul := λ x, quotient.induction_on' x $ λ _, congr_arg coe $ one_mul _ }
+
+variables {c}
+
+/-- The 1 of the quotient of a monoid by a congruence relation is the equivalence class of the
+    monoid's 1. -/
+@[simp, to_additive "The 0 of the quotient of an `add_monoid` by an additive congruence relation is the equivalence class of the `add_monoid`'s 0."]
+lemma coe_one : ((1 : M) : c.quotient) = 1 := rfl
+
+variables (M c)
+
+/-- The submonoid of `M × M` defined by a congruence relation on a monoid `M`. -/
+@[to_additive add_submonoid "The `add_submonoid` of `M × M` defined by an additive congruence relation on an `add_monoid` `M`."]
+protected def submonoid : submonoid (M × M) :=
+{ carrier := { x | c x.1 x.2 },
+  one_mem' := c.iseqv.1 1,
+  mul_mem' := λ _ _, c.mul }
+
+variables {M c}
+
+/-- The congruence relation on a monoid `M` from a submonoid of `M × M` for which membership
+    is an equivalence relation. -/
+@[to_additive of_add_submonoid "The additive congruence relation on an `add_monoid` `M` from an `add_submonoid` of `M × M` for which membership is an equivalence relation."]
+def of_submonoid (N : submonoid (M × M)) (H : equivalence (λ x y, (x, y) ∈ N)) : con M :=
+{ r := λ x y, (x, y) ∈ N,
+  iseqv := H,
+  mul' := λ _ _ _ _, N.mul_mem }
+
+/-- Coercion from a congruence relation `c` on a monoid `M` to the submonoid of `M × M` whose
+    elements are `(x, y)` such that `x` is related to `y` by `c`. -/
+@[to_additive to_add_submonoid "Coercion from a congruence relation `c` on an `add_monoid` `M` to the `add_submonoid` of `M × M` whose elements are `(x, y)` such that `x` is related to `y` by `c`."]
+instance to_submonoid : has_coe (con M) (submonoid (M × M)) := ⟨λ c, c.submonoid M⟩
+
+@[to_additive] lemma mem_coe {c : con M} {x y} :
+  (x, y) ∈ (↑c : submonoid (M × M)) ↔ (x, y) ∈ c := iff.rfl
+
+@[to_additive to_add_submonoid_inj]
+theorem to_submonoid_inj (c d : con M) (H : (c : submonoid (M × M)) = d) : c = d :=
+ext $ λ x y, show (x, y) ∈ (c : submonoid (M × M)) ↔ (x, y) ∈ ↑d, by rw H
+
+@[to_additive]
+lemma le_iff {c d : con M} : c ≤ d ↔ (c : submonoid (M × M)) ≤ d :=
+⟨λ h x, h x.1 x.2, λ h x y hc, h $ show (x, y) ∈ c, from hc⟩
+
+/-- The kernel of a monoid homomorphism as a congruence relation. -/
+@[to_additive "The kernel of an `add_monoid` homomorphism as an additive congruence relation."]
+def ker (f : M →* P) : con M := mul_ker f f.3
+
+/-- The definition of the congruence relation defined by a monoid homomorphism's kernel. -/
+@[to_additive "The definition of the additive congruence relation defined by an `add_monoid` homomorphism's kernel."]
+lemma ker_rel (f : M →* P) {x y} : ker f x y ↔ f x = f y := iff.rfl
+
+variables (c)
+
+/-- Restriction of a congruence relation on a monoid to a submonoid. -/
+@[to_additive "Restriction of a congruence relation on an `add_monoid` to an `add_submonoid`."]
+def subtype (A : submonoid M) : con A :=
+⟨λ x y, c x y, ⟨λ x, c.refl x, λ _ _, c.symm, λ _ _ _, c.trans⟩,
```

Comment:
This could just be `comap`

Reply:
I've generalised the 'second isomorphism theorem for monoids' a bit based on this and added a 'second isomorphism theorem for sets' but I'm not really sure there's much point to their existences.